File: TicTacToeGame.cpp
#include <deque>
#include <iostream>
#include <string>
#include <vector>

class TicTacToeGame {
private:
    Board* board;
    std::deque<TicTacToePlayer*> players;
    TicTacToeRules* rules;
    std::vector<IObserver*> observers;
    bool gameOver;

public:
    TicTacToeGame(int boardSize)
        : board(new Board(boardSize)), rules(new StandardTicTacToeRules()), gameOver(false) {}

    void addPlayer(TicTacToePlayer* player) {
        players.push_back(player);
    }

    void addObserver(IObserver* observer) {
        observers.push_back(observer);
    }

    void notify(std::string msg) {
        for(auto observer : observers) {
            observer->update(msg);
        }
    }

    void play() {
        if(players.size() < 2) {
            std::cout << "Need at least 2 players!" << std::endl;
            return;
        }

        notify("Tic Tac Toe Game Started!");

        while(!gameOver) {
            board->display();

            TicTacToePlayer* currentPlayer = players.front();
            std::cout << currentPlayer->getName() << " (" << currentPlayer->getSymbol()->getMark() << ") - Enter row and column: ";

            int row, col;
            std::cin >> row >> col;

            if(rules->isValidMove(board, row, col)) {
                board->placeMark(row, col, currentPlayer->getSymbol());
                notify(currentPlayer->getName() + " played (" + std::to_string(row) + "," + std::to_string(col) + ")");

                if(rules->checkWinCondition(board, currentPlayer->getSymbol())) {
                    board->display();
                    std::cout << currentPlayer->getName() << " wins!" << std::endl;
                    currentPlayer->incrementScore();
                    notify(currentPlayer->getName() + " wins!");
                    gameOver = true;
                }
                else if(rules->checkDrawCondition(board)) {
                    board->display();
                    std::cout << "It's a draw!" << std::endl;
                    notify("Game is Draw!");
                    gameOver = true;
                }
                else {
                    players.pop_front();
                    players.push_back(currentPlayer);
                }
            }
            else {
                std::cout << "Invalid move! Try again." << std::endl;
            }
        }
    }

    ~TicTacToeGame() {
        delete board;
        delete rules;
    }
};

File: frontend/src/logic/TicTacToeGame.js
import { Board } from './Board.js';
import { StandardTicTacToeRules } from './StandardTicTacToeRules.js';

export class TicTacToeGame {
    constructor(boardSize) {
        this.boardSize = boardSize;
        this.board = new Board(boardSize);
        this.players = [];
        this.rules = new StandardTicTacToeRules();
        this.gameOver = false;
        this.winner = null;
        this.isDraw = false;
        this.observers = [];
    }
    
    addObserver(observer) {
        this.observers.push(observer);
    }
    
    notify(msg) {
        for (const observer of this.observers) {
            observer.update(msg);
        }
    }

    addPlayer(player) {
        this.players.push(player);
    }

    getCurrentPlayer() {
        if (this.players.length === 0) return null;
        return this.players[0];
    }
    
    getBoard() {
        return this.board.getFlatBoard();
    }
    
    isGameOver() {
        return this.gameOver;
    }
    
    getWinner() {
        return this.winner;
    }
    
    getIsDraw() {
        return this.isDraw;
    }

    makeMove(row, col) {
        if (this.gameOver) return false;
        if (this.players.length < 2) return false;

        let currentPlayer = this.getCurrentPlayer();

        if (this.rules.isValidMove(this.board, row, col)) {
            this.board.placeMark(row, col, currentPlayer.getSymbol());
            this.notify(`${currentPlayer.getName()} played (${row},${col})`);

            if (this.rules.checkWinCondition(this.board, currentPlayer.getSymbol())) {
                currentPlayer.incrementScore();
                this.winner = currentPlayer;
                this.gameOver = true;
                this.notify(`${currentPlayer.getName()} wins!`);
            } else if (this.rules.checkDrawCondition(this.board)) {
                this.isDraw = true;
                this.gameOver = true;
                this.notify("Game is Draw!");
            } else {
                // switch turns
                this.players.shift();
                this.players.push(currentPlayer);
            }
            return true;
        }
        return false;
    }
}
